### PR TITLE
ci: add deploy workflow

### PR DESCRIPTION
**Pull request by Kyle-Verhoog**
This patch adds a deploy workflow for automatically deploying the package to PYPI when a release is "published".

According to the docs for "published":

> published: a release, pre-release, or draft of a release is published

[source](https://docs.github.com/en/developers/webhooks-and-events/webhook-events-and-payloads#release)

Unfortunately Github actions [doesn't have a manual approval step](https://github.community/t/github-actions-manual-trigger-approvals/16233/85) which would be ideal to use here.

This is based off of the [cibuildwheel example](https://github.com/joerick/cibuildwheel/blob/5c4767899dc4418b509a2d7ccfde07d33ab29e28/examples/github-deploy.yml) and our modifications made in #1554.

### CODE REVIEW

File: .github/workflows/deploy.yml [hunk 1]
```diff
@@ -0,0 +1,45 @@
...
+          fetch-depth: 0
+
+      - uses: actions/setup-python@v2
+        name: Install Python
+        with:
+          python-version: '3.8'
+
+      - name: Install dependencies
+        run: |
+          python -m pip install cibuildwheel==1.5.1 setuptools wheel twine
```
Comment (jd) [inline .github/workflows/deploy.yml:28]: any good reason to fix the version?
we might forget to update for long

Comment (Kyle-Verhoog) [reply .github/workflows/deploy.yml:28]: No - I think it's just what they include in their [examples](https://github.com/joerick/cibuildwheel/tree/master/examples).

It shouldn't matter if we update for long though and I think we'd prefer to have reproducible runs as much as possible here.

**Comment by brettlangdon** [comment]
Do we have a way to validate this before making it do the real thing?

**Comment by Kyle-Verhoog** [comment]
Unfortunately no easy way that I know of. I was thinking of creating another repo and using the test pypi instance.

**Comment by brettlangdon** [comment]
good luck

**Comment by Kyle-Verhoog** [comment]
I am proceeding with doing a test run with https://github.com/Kyle-Verhoog/dd-trace-py-test.

**Comment by Kyle-Verhoog** [comment]
It seems to work (besides the auth failure): https://github.com/Kyle-Verhoog/dd-trace-py-test/runs/950901623?check_suite_focus=true.

**Pull request merged**